## Summary

This package allows to keep track of activities, gains, and losses of a
investment portfolio. Use commands to log purchase and sales of shares,
dividends, and to update the value of the portfolio.

The basic feature of the package is that it can produce tables containing
the relative value of each share. This takes into account the purchase date,
purchase price and fee, dividends that were paid, the current value of the
shares, the sales fee, and the number of days for which the investment was
held. A relative value of p means that the investment is equivalent to having
put the purchase amount into a savings account with continously compounded
interest of p percent per year.

A number of additional tools are provided, e.g. a method that computes bond
prices. The significance of p values can be understood by analysing historic
data. The notebook <code>ta_work</code> can be used for working on the
account -- it contains a list of all callable functions (for which the
docstrings can be displayed).

## Initialisation

To demonstrate the trading account (TA) logbook package,
we first execute the master file. This will load all necessary
python packages and all the TA classes and methods.

In [1]:
# execute master script
master_file = open("./ta_master.py")
exec(master_file.read())
master_file.close()

If no files containing TA data etc. are found, new ones will be
initiated. Several TA logbooks can be kept in parallel. We can
check which ones are present by running the following command.

In [2]:
account_name()

['demo', 'ta']

If the working directory was clean, the produced list will only
contain the default account <code>ta</code>. We can add a new one by
passing its name to the function <code>account_name()</code>. We now create
a demo account for this demonstration. If <code>demo</code> is already
present, the command will simply switch to it.

In [3]:
account_name('demo')

Switched to existing account, demo.


['demo', 'ta']

In this list of accounts, the first entry is always the active
account. The functionalities of the TA package will be
demonstrated in this notebook, but we can always have the
docstrings displayed for further details, e.g.:

In [4]:
print(account_name.__doc__)

Display current account name, switch to others, or create new one.

    Optional arguments:
    acct_name -- name of account (string) to switch to (if name exists)
                                    or to create (if name does not exit)

    Note:
    Return list of existing accounts with active/current
                                    one in the first position.
    


The callable methods are all listed in the <code>ta_work</code> notebook.
The logbooks for each files are pandas dataframes, that are saved
in the working directory. Once the software is initialised, we
can simply run the commands described below to act on the active
TA (e.g., use <code>ta_work</code>) for that. Before doing that, we
reset the <code>demo</code> TA:

In [5]:
if os.path.isfile('demo_save.p'):
    os.remove('demo_save.p')

## Logging TA Activities

We first deposit 5000 in the account, then add and withdraw, and make a small adjustment.

In [6]:
account_activity(5000, date=pd.Timestamp(2017, 1, 1))
account_activity(1000, date=pd.Timestamp(2017, 2, 1))
account_activity(-500, date=pd.Timestamp(2017, 3, 1))
account_activity(5, comment='Adjustment', date=pd.Timestamp(2017, 4, 1))

No trading account log file found, created new one.


A new TA file with the name <code>demo</code> with an opening deposit of 5000 was created. If <code>date</code> is not given, the current date will be used -- this is what one would do in practice (in this demo, dates are prescribed to demonstrate the changes in the account over a period of a few months). We can now display the account:

In [7]:
all_values()

Date  Acct Bal          Comment
0 2017-01-01      5000  Opening deposit
1 2017-02-01      6000          Deposit
2 2017-03-01      5500       Withdrawal
3 2017-04-01      5505       Adjustment

The different display options will be discussed in the next section. Comments are set automaticall, but the default comments can be overwritten, cf. the docstring for <code>account_activity()</code>.  First, let us by some shares, namely 2000 worth of <code>Share1</code>, with an order fee of 10.

In [8]:
buy('Share1', 2000, 10, date=pd.Timestamp(2017, 5, 1))
all_values()

Date  Acct Bal          Comment               Share1
0 2017-01-01      5000  Opening deposit     nan (0.00, 0.00)
1 2017-02-01      6000          Deposit     nan (0.00, 0.00)
2 2017-03-01      5500       Withdrawal     nan (0.00, 0.00)
3 2017-04-01      5505       Adjustment     nan (0.00, 0.00)
4 2017-05-01      3495       Buy Share1  nan (2000.00, 0.00)

The value of the shares plus the order fee is automatically subtracted from the account balance. Before interpreting the displayed values, let us update the share values: suppose that after 1 month, the value of the share has increased by 50:

In [9]:
update(Share1=2050, date=pd.Timestamp(2017, 6, 1))
all_values()

Date  Acct Bal          Comment                  Share1
0 2017-01-01      5000  Opening deposit        nan (0.00, 0.00)
1 2017-02-01      6000          Deposit        nan (0.00, 0.00)
2 2017-03-01      5500       Withdrawal        nan (0.00, 0.00)
3 2017-04-01      5505       Adjustment        nan (0.00, 0.00)
4 2017-05-01      3495       Buy Share1     nan (2000.00, 0.00)
5 2017-06-01      3495           Update  0.1455 (2050.00, 0.00)

The first argument in parentheses is the share value. The second value is the total amount of dividends paid for this share -- 0 so far. The value in front of the parentheses, 0.1744, is the relative share value -- having this information in the TA logbook was the main motivation for writing this package. It gives the continously compounded interest rate necessary to make the same profit over the given time period:

costs, on 05-01:  2010,

payout, on 06-01:  2040   (the share value minus an estimated order fee of 10),

and to turn 2010 into 2040 within one month, we would need a savings account with a continuously compounded interest rate of 17.44%:

In [10]:
2010 * math.exp( float(31)/365 * 0.1744)

2039.9938167885055

Now, whether or not this is high enough to encourage us to sell the share is a different story -- it depends on the properties of the share, the market, and our expectations. Tools to gauge the significance of relative share values such as 0.1744 are included in the package and will be discussed later. Had there been any dividend payments for the share, then they would be taken into account for the relative value. Next, buy another share and update share values:

In [11]:
update(Share1=2050, date=pd.Timestamp(2017, 7, 1))
buy('Share2', 1000, 10, date=pd.Timestamp(2017, 7, 1))
update(Share1=2040, Share2=1020, date=pd.Timestamp(2017, 8, 1))
all_values()

Date  Acct Bal          Comment                  Share1  \
0 2017-01-01      5000  Opening deposit        nan (0.00, 0.00)   
1 2017-02-01      6000          Deposit        nan (0.00, 0.00)   
2 2017-03-01      5500       Withdrawal        nan (0.00, 0.00)   
3 2017-04-01      5505       Adjustment        nan (0.00, 0.00)   
4 2017-05-01      3495       Buy Share1     nan (2000.00, 0.00)   
5 2017-06-01      3495           Update  0.1455 (2050.00, 0.00)   
6 2017-07-01      3495           Update  0.0740 (2050.00, 0.00)   
7 2017-07-01      2485       Buy Share2  0.0740 (2050.00, 0.00)   
8 2017-08-01      2485           Update  0.0295 (2040.00, 0.00)   

                    Share2  
0         nan (0.00, 0.00)  
1         nan (0.00, 0.00)  
2         nan (0.00, 0.00)  
3         nan (0.00, 0.00)  
4         nan (0.00, 0.00)  
5         nan (0.00, 0.00)  
6         nan (0.00, 0.00)  
7      nan (1000.00, 0.00)  
8  -0.0584 (1020.00, 0.00)

For <code>update()</code>, the current share values are given as a dictionary with the share names as keys. If those values are not given, only the relative share values will be updated according to their time-dependancy. Note how the relative value of <code>Share1</code> has decreased in row 6 despite the share value remaining the same -- that's because more time has passed. Next we log a dividend payment and then sell one of the shares.

In [12]:
dividend('Share1', 200, date=pd.Timestamp(2017, 9, 1))
update(Share1=2070, Share2=1050, date=pd.Timestamp(2017, 10, 1))
all_values()

Date  Acct Bal          Comment                    Share1  \
0  2017-01-01      5000  Opening deposit          nan (0.00, 0.00)   
1  2017-02-01      6000          Deposit          nan (0.00, 0.00)   
2  2017-03-01      5500       Withdrawal          nan (0.00, 0.00)   
3  2017-04-01      5505       Adjustment          nan (0.00, 0.00)   
4  2017-05-01      3495       Buy Share1       nan (2000.00, 0.00)   
5  2017-06-01      3495           Update    0.1455 (2050.00, 0.00)   
6  2017-07-01      3495           Update    0.0740 (2050.00, 0.00)   
7  2017-07-01      2485       Buy Share2    0.0740 (2050.00, 0.00)   
8  2017-08-01      2485           Update    0.0295 (2040.00, 0.00)   
9  2017-09-01      2685  Dividend Share1  0.3016 (2040.00, 200.00)   
10 2017-10-01      2685           Update  0.2744 (2070.00, 200.00)   

                     Share2  
0          nan (0.00, 0.00)  
1          nan (0.00, 0.00)  
2          nan (0.00, 0.00)  
3          nan (0.00, 0.00)  
4          nan (0.00, 0.00)  
5          nan (0.00, 0.00)  
6          nan (0.00, 0.00)  
7       nan (1000.00, 0.00)  
8   -0.0584 (1020.00, 0.00)  
9   -0.0584 (1020.00, 0.00)  
10   0.0970 (1050.00, 0.00)

In [13]:
sell('Share1',2065,date = pd.Timestamp(2017,10,1))
update(Share2=1080,date = pd.Timestamp(2017,11,1))
all_values()

Share1 was sold with an overall return of 28.5%.


Date  Acct Bal          Comment                   Share2
0  2017-01-01      5000  Opening deposit         nan (0.00, 0.00)
1  2017-02-01      6000          Deposit         nan (0.00, 0.00)
2  2017-03-01      5500       Withdrawal         nan (0.00, 0.00)
3  2017-04-01      5505       Adjustment         nan (0.00, 0.00)
4  2017-05-01      3495       Buy Share1         nan (0.00, 0.00)
5  2017-06-01      3495           Update         nan (0.00, 0.00)
6  2017-07-01      3495           Update         nan (0.00, 0.00)
7  2017-07-01      2485       Buy Share2      nan (1000.00, 0.00)
8  2017-08-01      2485           Update  -0.0584 (1020.00, 0.00)
9  2017-09-01      2685  Dividend Share1  -0.0584 (1020.00, 0.00)
10 2017-10-01      2685           Update   0.0970 (1050.00, 0.00)
11 2017-10-01      4750      Sell Share1   0.0970 (1050.00, 0.00)
12 2017-11-01      4750           Update   0.1573 (1080.00, 0.00)

When selling, one would simply pass the total amount credited to the account as an argument. E.g., here, the share value of 2070 minus a 5 order fee. Upon selling, a message is produced stating the overall relative return (here, this return is a little bit better than the last relative value that was displayed, because the order fee for the sale turned out to be less than the 10 that are used for computing the relative values; the default sales fee can be set in the code). Note that the values for <code>Share1</code> are not displayed, because it is not an active share anymore, cf. next section. If a mistake is made handling the account, it can be undone as follows.

In [14]:
account_activity(-3, comment='Mistake', date=pd.Timestamp(2017, 11, 1))
all_values()

Date  Acct Bal          Comment                   Share2
0  2017-01-01      5000  Opening deposit         nan (0.00, 0.00)
1  2017-02-01      6000          Deposit         nan (0.00, 0.00)
2  2017-03-01      5500       Withdrawal         nan (0.00, 0.00)
3  2017-04-01      5505       Adjustment         nan (0.00, 0.00)
4  2017-05-01      3495       Buy Share1         nan (0.00, 0.00)
5  2017-06-01      3495           Update         nan (0.00, 0.00)
6  2017-07-01      3495           Update         nan (0.00, 0.00)
7  2017-07-01      2485       Buy Share2      nan (1000.00, 0.00)
8  2017-08-01      2485           Update  -0.0584 (1020.00, 0.00)
9  2017-09-01      2685  Dividend Share1  -0.0584 (1020.00, 0.00)
10 2017-10-01      2685           Update   0.0970 (1050.00, 0.00)
11 2017-10-01      4750      Sell Share1   0.0970 (1050.00, 0.00)
12 2017-11-01      4750           Update   0.1573 (1080.00, 0.00)
13 2017-11-01      4747          Mistake   0.1573 (1080.00, 0.00)

In [15]:
delete_last_row()
all_values()

Backed up trading account and deleted last row.


Date  Acct Bal          Comment                   Share2
0  2017-01-01      5000  Opening deposit         nan (0.00, 0.00)
1  2017-02-01      6000          Deposit         nan (0.00, 0.00)
2  2017-03-01      5500       Withdrawal         nan (0.00, 0.00)
3  2017-04-01      5505       Adjustment         nan (0.00, 0.00)
4  2017-05-01      3495       Buy Share1         nan (0.00, 0.00)
5  2017-06-01      3495           Update         nan (0.00, 0.00)
6  2017-07-01      3495           Update         nan (0.00, 0.00)
7  2017-07-01      2485       Buy Share2      nan (1000.00, 0.00)
8  2017-08-01      2485           Update  -0.0584 (1020.00, 0.00)
9  2017-09-01      2685  Dividend Share1  -0.0584 (1020.00, 0.00)
10 2017-10-01      2685           Update   0.0970 (1050.00, 0.00)
11 2017-10-01      4750      Sell Share1   0.0970 (1050.00, 0.00)
12 2017-11-01      4750           Update   0.1573 (1080.00, 0.00)

Other corrections have to be done manually. The correction method <code>delete_last_row()</code> also calls the function <code>backup()</code>, which should be run from time to time. It saves the dataframe, both as a python object and as a spreadheet (with a timestamp). 

## Dislaying the TA Log

In the displayed TA dataframes above, the column for <code>Share1</code> was dropped after it had been sold. It is still in the dataframe, of course, and it can be displayed as follows.

In [16]:
all_values(all_shares=True)

Date  Acct Bal          Comment                    Share1  \
0  2017-01-01      5000  Opening deposit          nan (0.00, 0.00)   
1  2017-02-01      6000          Deposit          nan (0.00, 0.00)   
2  2017-03-01      5500       Withdrawal          nan (0.00, 0.00)   
3  2017-04-01      5505       Adjustment          nan (0.00, 0.00)   
4  2017-05-01      3495       Buy Share1       nan (2000.00, 0.00)   
5  2017-06-01      3495           Update    0.1455 (2050.00, 0.00)   
6  2017-07-01      3495           Update    0.0740 (2050.00, 0.00)   
7  2017-07-01      2485       Buy Share2    0.0740 (2050.00, 0.00)   
8  2017-08-01      2485           Update    0.0295 (2040.00, 0.00)   
9  2017-09-01      2685  Dividend Share1  0.3016 (2040.00, 200.00)   
10 2017-10-01      2685           Update  0.2744 (2070.00, 200.00)   
11 2017-10-01      4750      Sell Share1          nan (0.00, 0.00)   
12 2017-11-01      4750           Update          nan (0.00, 0.00)   

                     Share2  
0          nan (0.00, 0.00)  
1          nan (0.00, 0.00)  
2          nan (0.00, 0.00)  
3          nan (0.00, 0.00)  
4          nan (0.00, 0.00)  
5          nan (0.00, 0.00)  
6          nan (0.00, 0.00)  
7       nan (1000.00, 0.00)  
8   -0.0584 (1020.00, 0.00)  
9   -0.0584 (1020.00, 0.00)  
10   0.0970 (1050.00, 0.00)  
11   0.0970 (1050.00, 0.00)  
12   0.1573 (1080.00, 0.00)

Now, the entries in the columns with share values are strings. We can also generate dataframes with only the share values or the relative values as floats.

In [17]:
df = shr_values(all_shares=True)
df

Acct Bal  Share1  Share2
Date                                
2017-01-01      5000       0       0
2017-02-01      6000       0       0
2017-03-01      5500       0       0
2017-04-01      5505       0       0
2017-05-01      3495    2000       0
2017-06-01      3495    2050       0
2017-07-01      3495    2050       0
2017-07-01      2485    2050    1000
2017-08-01      2485    2040    1020
2017-09-01      2685    2040    1020
2017-10-01      2685    2070    1050
2017-10-01      4750       0    1050
2017-11-01      4750       0    1080

Use <code>rel_values()</code> to produce a dataframe showing the relative share values. Note that the data was reindexed over time, so that it can be plotted.

In [18]:
df.plot()

The three different display methods <code>all_values(), rel_values(), shr_values()</code> have different default formats, but this can be overwritten with keyword arguments, e.g. <code>shr_values(comments=True, date_as_index=False)</code>. Cf. the docstrings, e.g.:

In [19]:
print(shr_values.__doc__)

Return dataframe with shares values as floats.

    Keyword arguments:
    all_shares -- display all shares instead of
                                    active ones only (default False)
    comments -- display comments (default False)
    acct_bal -- display account balance (default True)
    date_as_string -- write dates as strings (default False)
    date_as_index -- set dates as index (default True)
    


## Other Methods on the Dataframe

List the shares in the portfolio using <code>active_shares()</code> or <code>all_shares()</code>. The methods <code>delete_last_row()</code> and <code>backup()</code> have already been discussed. Lastly, the total value of the portfolio:

In [20]:
total_value()

Total Value
Date                   
2017-01-01         5000
2017-02-01         6000
2017-03-01         5500
2017-04-01         5505
2017-05-01         5480
2017-06-01         5530
2017-07-01         5530
2017-07-01         5505
2017-08-01         5515
2017-09-01         5715
2017-10-01         5775
2017-10-01         5785
2017-11-01         5815

## Investment tools

Suppose we are considering to buy a certain bond with a coupon (yield) of 4.3% and 15 years to maturity. Then we can use the following function to get an overview of bond prices and the corresponding overall return.

In [21]:
bond_evaluation(4.3, 15)

Price
Return (%)            
-1.0        186.237255
-0.5        174.963471
 0.0        164.500000
 0.5        154.783174
 1.0        145.754673
 1.5        137.361052
 2.0        129.553306
 2.5        122.286480
 3.0        115.519316
 3.5        109.213929
 4.0        103.335516
 4.5         97.852091
 5.0         92.734239
 5.5         87.954903
 6.0         83.489177
 6.5         79.314129
 7.0         75.408632
 7.5         71.753217
 8.0         68.329929
 8.5         65.122206
 9.0         62.114764
 9.5         59.293490

This means that, if we want a return of at least 5%, we should not pay more than 92.73. This is an unrelated functionality -- the following methods are more geared towards the main features of the package described above.

The following methods help gauge the significance of relative values p -- this helps to plan ahead and also to decide whether to hold or to sell a share. For example, if a share with very low volatility has a p-value double its average historical rate of return, then this is a exceptional event, and we should sell before the share price reverts to the mean. For more volatile shares, such a p-value is not as significant and we may consider holding.

To do this, download weekly closing prices and save it in the working directory as <code>data.csv</code>. Then run the following to find the average weekly growth and its standard deviation from that mean (alternatively, a list of weekly closing prices can be passed).

In [22]:
parameters = find_mu_sigma()
parameters

[0.003671379220282967, 0.03903976147063121]

Next, we use this data for a Monte-Carlo simulation of shares with those parameter. This produces a distribution of maximum achieved relative rates.

In [23]:
simulate_p(parameters[0], parameters[1])

Date Sharename   p_max    p_90    p_80    p_70    p_60    p_50    p_40  \
0  19-02-17            2.7563  1.1655  0.9398  0.7886  0.6693  0.5502  0.4448   
1                          12      14      12      18      24      44      21   

     p_30    p_20    p_10    p_min  
0  0.3340  0.2136  0.0606  -0.7190  
1      52      37      52       24

The numbers in the second row state the weeks in which those p-values were achieved (only p values between certain weeks are considered, cf. the docstring for <code>simulate_p()</code>).

Let's check that a lower volatility gives a more concentrated distribution of p-values.

In [24]:
simulate_p(parameters[0], 0.001*parameters[1], name='demoshare')

Date  Sharename   p_max    p_90    p_80    p_70    p_60    p_50  \
0  19-02-17  demoshare  0.2106  0.2091  0.2088  0.2086  0.2084  0.2083   
1                           12      12      12      12      12      12   

     p_40    p_30    p_20    p_10   p_min  
0  0.2081  0.2080  0.2077  0.2075  0.2060  
1      12      12      12      12      12

This time, the name of the share was given as an keyword argument. In this case, the above data is written in a spreadsheet in the working directory. One can later consult this spreadsheet when making decisions.

Lastly, as a sanity check, let us make sure that the median p found in the low-volatility simulation really corresponds to the average weekly rate found above:

In [25]:
p_med = 0.2308 # = p_50 in the previous table
               # change if you were running this on different data
               # should approximately give parameters[0]
math.exp(p_med)**(float(1)/52)

1.0044483260979653

## Web Scraping Functionalities

Suppose there are stocks we'd like to buy, but they are currently overpriced. We can keep a list of stocks, <code>watchlist.txt</code>, with the Yahoo Finance handles and the stock prices at which we would buy. Calling the following function will check the watchlist (this is a separate script, so, it can be run from outside the whole ta package). This is very helpful once that list gets very long.

In [26]:
check_watchlist()


ALERT!!!
KO is currently 45.240 < 100.000.



Next, we use the same type of scraping to produce the list of arguments that is fed into the method <code>update()</code> -- this will save a lot of time, especially if the portfolio is big. For this, a text file with the name <code>'demo_dict.txt'</code> needs to be kept in the working directory (replace  <code>'demo'</code> with the name of the account). Each line of that file contains the name of the stock in the TA logbook (here, <code>'Share1'</code> and <code>'Share2'</code>), the Yahoo Finance handle, and the number of stock (shares/bonds/certificates). In a later version of this software package, the YF handle and number of stock should be included in the objects that keep the data, so that this separate dictionary file becomes unnecessary. Well, for now we go via the dictionaries -- here is how it works.

In [27]:
ud = get_update_dict()
print(ud)

{'Share1': 1787.5, 'Share2': 904.8000000000001}


This can now be fed into the methods <code>update()</code>:

In [28]:
update(**ud)
df = shr_values(all_shares=True)
df

Acct Bal  Share1 Share2
Date                                               
2017-01-01 00:00:00.000000      5000       0      0
2017-02-01 00:00:00.000000      6000       0      0
2017-03-01 00:00:00.000000      5500       0      0
2017-04-01 00:00:00.000000      5505       0      0
2017-05-01 00:00:00.000000      3495    2000      0
2017-06-01 00:00:00.000000      3495    2050      0
2017-07-01 00:00:00.000000      3495    2050      0
2017-07-01 00:00:00.000000      2485    2050   1000
2017-08-01 00:00:00.000000      2485    2040   1020
2017-09-01 00:00:00.000000      2685    2040   1020
2017-10-01 00:00:00.000000      2685    2070   1050
2017-10-01 00:00:00.000000      4750       0   1050
2017-11-01 00:00:00.000000      4750       0   1080
2019-02-17 15:04:31.988045      4750       0  904.8

Note how the first share was not updated -- that's because it is not active anymore. Hence the stock dictionary does not need to be updated -- well, only if new stocks are purchased. If a an inactive share is re-purchased, it needs to be given a new name (this is not only required for the steps surrounding the web scraping functionality, but also for other parts of the package; this is intentional, i.e. not a bug). One case use the method <code>auto_update()</code> to combine these two steps (scrape the value of the portfolio and update it).

## Getting Started

Now that you know how everything works, just run

<code>account_name('your_account_name')</code>

and start managing your account! This can be done in another notebook (e.g. <code>ta_work</code>) or from the python command line -- just execute <code>ta_master.py</code> first to load everything.

## To-Do

Other functionalities that could be implemented: use scraped data for finding the significance levels and for the fundamental analysis; include Yahoo Finance handle and number of stock in the share value objects (so that the stock dictionary is not needed anymore); for fitting stock data to obtain parameters for the Monte Carlo simulation, use MLE with normal inverse Gaussian distribution rather than normal distributions.

Please email me with any questions, concerns, bugs, suggestions.